### PR TITLE
Proposal: edits to RELEASE_GUIDE.md

### PR DESCRIPTION
**Pull request by JordanMartinez**
**Description of the change**

Continuing a discussion in #4102.

---

**Checklist:**

- [ ] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [ ] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

### CODE REVIEW

File: RELEASE_GUIDE.md
Comment (JordanMartinez) [inline RELEASE_GUIDE.md:22]: Note: these two `stack exec` commands don't work. I'm not sure how to pass the source globs correctly to the resulting `purs` binary.

File: RELEASE_GUIDE.md
Comment (JordanMartinez) [inline RELEASE_GUIDE.md:160]: Would this be 
```bash
# if Linux
CI_RELEASE="true" ./bundle/build.sh "Linux"
# if Mac
CI_RELEASE="true" ./bundle/build.sh "macOS"
# if Windows
CI_RELEASE="true" ./bundle/build.sh "Windows"
```

**Comment by JordanMartinez** [comment]
Should [these tools](https://github.com/purescript/purescript/pull/4104/files#diff-3dd37fe82be5315d56d3db6442ffe1e6993df584f98cc13d3aff2f02a027843eR56) have links to their respective repos?

Let's say there are issues with [these JSON formats](https://github.com/purescript/purescript/pull/4104/files#diff-3dd37fe82be5315d56d3db6442ffe1e6993df584f98cc13d3aff2f02a027843eR63-R66). What do we do?

File: RELEASE_GUIDE.md
Comment (f-f) [reply RELEASE_GUIDE.md:22]: If you run `stack exec -- bash` then you'll be dropped into a shell, so maybe that'd be easier?

File: RELEASE_GUIDE.md
Comment (f-f) [inline RELEASE_GUIDE.md:19]: We really need a `spago install --whole-set` 😄

**Review by f-f — COMMENTED** [review]
Thank you for writing these up!

File: RELEASE_GUIDE.md
Comment (hdgarrood) [reply RELEASE_GUIDE.md:22]: Yeah, I think `stack exec bash` followed by `spago build` and `spago docs` is probably easiest.

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:70]: Can you avoid the passive voice for actions that the maintainer needs to take please? I think it is making it unclear which steps are performed automatically and which ones need to be done by the person making the release. Perhaps something like: "To make a release, a maintainer should create a PR updating versions in the repo, and once that is merged, they can create a GitHub release, which triggers a CI build ..."

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:103]: I think "Finalizing CHANGELOG for {version}" might be a slightly clearer commit message, if we are going to suggest one here.

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:97]:
```suggestion
  #   until satisfied (e.g. grammar,
```

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:150]: I am not sure that the changes to this section are an improvement. They add quite a bit of noise, and explain how to do basic tasks like creating git branches, opening files in  their editor, or submitting PRs, but we should assume that a maintainer creating a release knows how to do these things. I think the extra noise is obscuring the important information here, which is which parts of which files need updating.

File: RELEASE_GUIDE.md
Comment (hdgarrood) [reply RELEASE_GUIDE.md:160]: It would be along those lines, but you'll also need to make sure you've built a release version of the compiler first as well (so that `purs --version` doesn't say it's a development build). In the name of saving effort, since this scenario might not even happen, I'm tempted to say let's not bother writing instructions. Since we are the maintainers of the CI workflow/release script, we should be able to figure out how to run it if need be.

**Comment by hdgarrood** [comment]
> Should these tools have links to their respective repos?

I don't see why not.

> Let's say there are issues with these JSON formats. What do we do?

There's no single answer to this, it depends on context. If a format has changed accidentally, then we should restore the old format before making the release. If a format has changed deliberately, we should make sure that we do a major version bump and that the change is noted in the breaking changes section of the changelog, and give the maintainers of any affected tools a warning in advance. Ideally all of this would happen at the time a PR is merged rather than when trying to cut a release though. We do have some backwards-compatibility JSON tests, but I don't think they're comprehensive right now.

File: RELEASE_GUIDE.md
Comment (JordanMartinez) [reply RELEASE_GUIDE.md:150]: Yeah... I have to agree with you here. It also doesn't account for people who are using editors that can easily navigate to such files.

File: RELEASE_GUIDE.md
Comment (JordanMartinez) [reply RELEASE_GUIDE.md:22]: I tried `stack exec bash` and it didn't drop into a shell. Found [this SO answer](https://stackoverflow.com/a/48014385) and think we should use this:
```bash
stack exec bash <<EOF
spago build
spago docs
EOF
```

**Comment by JordanMartinez** [comment]
Following the state of the RELEASE_GUIDE.md as it is currently stated in this PR, I've...
- confirmed that the latest package set compiles with this version
- unsure whether and what needs to change in `INSTALL.md`, but I believe this can be resolved in a separate PR if needed.
- verified (by looking at the changelog) that there are no breaking changes
- merged the 'version bump and changelog finalization' PR via #4103

I'll now trigger the GH-release CI build. I still can't upload to Hackage yet, but should be able to update the npm package.

File: RELEASE_GUIDE.md
Comment (hdgarrood) [reply RELEASE_GUIDE.md:21]: Is this just `spago verify-set`?

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:77]: You've mentioned this both at the start and the end of this section. I don't think it's necessary to mention it twice; can we remove one?

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:122]: This item and all of the other items following it need to be dedented so they aren't sub-items of the "If purescript-cst has changed" part, because all of these things need to happen every time, regardless of whether purescript-cst has changed.

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:170]:
```suggestion
stack upload .
```

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:166]: This doesn't work without the `.`
```suggestion
stack upload .
```

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:24]: You probably don't want the first `exec` here. It will make your terminal/shell exit after the command is complete.
```suggestion
stack exec bash <<EOF
```

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:51]: Typo here? "Open a meta issue in purescript/purescript issue" doesn't make sense to me. I'm also not sure we need to repeat the whole "open a "meta" issue" spiel for every item in this list.

```suggestion
- core libraries
- contrib libraries
- node libraries
- web libraries

It may be helpful to open a "meta" issue in the relevant repository to track progress for each group of libraries which need updating. For example, [purescript/purescript#3942](https://github.com/purescript/purescript/issues/3942).
```

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:83]: Let's get rid of this too

File: RELEASE_GUIDE.md
Comment (hdgarrood) [inline RELEASE_GUIDE.md:178]: If we're using cabal, then there's no need to do this in a separate directory, because `cabal install xyz` installs the package globally and doesn't do anything with the current working directory. But I think we should avoid adding separate instructions here anyway, because we should be using the same instructions as we provide in INSTALL.md for installing the compiler from source.


```suggestion
# 3. Verify that `purescript-cst` and `purescript` can be installed from Hackage by following the instructions in INSTALL.md for building the compiler from source.
```

**Comment by hdgarrood** [comment]
Actually, maybe I'm going about reviewing this wrong. I think there are some changes in here which are good, but I also think that many of the places you've changed are better off as they are at the moment. So maybe I should just list the bits that I think are good and should go in, and we can just get those merged?

- :+1: on the prerequisites section
- :+1: on adding instructions for checking that there are no unintended breaking changes by building the package set
- :+1: on the clarification on how to check that INSTALL.md is up to date
- :+1: on linking to the tools

But I don't think the "Make a release" section is improved by these changes, so I'd prefer to leave it as-is.

File: RELEASE_GUIDE.md
Comment (JordanMartinez) [reply RELEASE_GUIDE.md:24]: Whoops!

**Comment by JordanMartinez** [comment]
@hdgarrood I've updated this PR to only include the :+1: changes from your previous comment. I'd rather get some of these in even if not all of them were accepted.

We may disagree philosophically on what this file should contain, and that's ok! Even if "maintainers should know that," my goal was to provide a checklist to prevent human error. Personally, I like procedures that tell me exactly what to do rather than leaving room for interpretation. If I have to spend 5 minutes interpreting what I'm supposed to do or remember which commands to run in the terminal, I just wasted 5 minutes. However, I realize that some of the changes I proposed above also specify the exact workflow to use. That's not good either because the approach I proposed here now may not be the one another uses, or even by the one I use later on.

On another hand, this PR may have been an overreaction on my part. Since I haven't made a PureScript release before, everything is new. Now that I have made one, I see how it works and some of the changes I proposed here might not be needed.

File: RELEASE_GUIDE.md
Comment (JordanMartinez) [reply RELEASE_GUIDE.md:21]: I think it would replace the `spago install` and `spago build` command. However, I am not sure whether it also builds docs. 

@f-f, could you clarify?

File: RELEASE_GUIDE.md
Comment (JordanMartinez) [reply RELEASE_GUIDE.md]: This was fixed in latest commits.

**Review by hdgarrood — APPROVED** [review]

File: RELEASE_GUIDE.md
Comment (f-f) [reply RELEASE_GUIDE.md:21]: The question "can I install the entirety of the set?" seems to come up often, and I don't have a good answer for that right now (it will improve once we get the new config format for Spago).

In this thread however, I think Harry is right: installing the whole set and then building it is _basically_ `verify-set` (which does some more stuff, so it's going to be slightly slower than just installing everything and building it, but it's fine)

**Pull request merged**
